File: test/bmp/BmpTest.cpp
////////////////////////////////////////////////////////////////////////
// Copyright (C) 2019 Xueyi Yao
//
// This file is part of VPixels.
//
// VPixels is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// VPixels is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with VPixels.  If not, see <https://www.gnu.org/licenses/>.
////////////////////////////////////////////////////////////////////////

#include "BmpTest.h"
#include "Bmp.h"
#include "Exception.h"

CPPUNIT_TEST_SUITE_REGISTRATION( BmpTest );

void BmpTest::testCtors()
{
  uint8_t B, G, R;

  // ctor with default values
  vp::Bmp bmp1;
  CPPUNIT_ASSERT( bmp1 );
  CPPUNIT_ASSERT( bmp1.BitsPerPixel() == 1 );
  CPPUNIT_ASSERT( bmp1.Width() == 1 );
  CPPUNIT_ASSERT( bmp1.Height() == 1 );
  // one pixel
  bmp1.SetPixel( 0, 0, 1 );
  CPPUNIT_ASSERT( bmp1.GetPixel( 0, 0 ) == 1 );
  CPPUNIT_ASSERT_THROW( bmp1.SetPixel( 0, 0, 23, 24, 25 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp1.GetPixel( 0, 0, B, G, R ), vp::Exception );
  // color table
  CPPUNIT_ASSERT( bmp1.ColorTableSize() == 2 );
  bmp1.SetColorTable( 0, 23, 24, 25 );
  bmp1.GetColorTable( 0, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );

  // ctor
  vp::Bmp bmp2{ 4, 10, 10 };
  CPPUNIT_ASSERT( bmp2 );
  CPPUNIT_ASSERT( bmp2.BitsPerPixel() == 4 );
  CPPUNIT_ASSERT( bmp2.Width() == 10 );
  CPPUNIT_ASSERT( bmp2.Height() == 10 );
  CPPUNIT_ASSERT( bmp2.ColorTableSize() == 16 );

  // copy ctor
  vp::Bmp bmp3{bmp1};
  CPPUNIT_ASSERT( bmp1 );
  CPPUNIT_ASSERT( bmp3 );
  CPPUNIT_ASSERT( bmp3.BitsPerPixel() == 1 );
  CPPUNIT_ASSERT( bmp3.Width() == 1 );
  CPPUNIT_ASSERT( bmp3.Height() == 1 );
  CPPUNIT_ASSERT( bmp3.ColorTableSize() == 2 );

  // copy assign
  bmp1 = bmp2;
  CPPUNIT_ASSERT( bmp1 );
  CPPUNIT_ASSERT( bmp2 );
  CPPUNIT_ASSERT( bmp1.BitsPerPixel() == 4 );
  CPPUNIT_ASSERT( bmp1.Width() == 10 );
  CPPUNIT_ASSERT( bmp1.Height() == 10 );
  CPPUNIT_ASSERT( bmp1.ColorTableSize() == 16 );

  // move ctor
  vp::Bmp bmp4{std::move(bmp2)};
  CPPUNIT_ASSERT( !bmp2 ); // bmp2 moved
  CPPUNIT_ASSERT( bmp4 );
  CPPUNIT_ASSERT( bmp4.BitsPerPixel() == 4 );
  CPPUNIT_ASSERT( bmp4.Width() == 10 );
  CPPUNIT_ASSERT( bmp4.Height() == 10 );
  CPPUNIT_ASSERT( bmp4.ColorTableSize() == 16 );

  // move assign
  bmp2 = std::move(bmp3);
  CPPUNIT_ASSERT( !bmp3 ); // bmp3 moved
  CPPUNIT_ASSERT( bmp2 );
  CPPUNIT_ASSERT( bmp2.BitsPerPixel() == 1 );
  CPPUNIT_ASSERT( bmp2.Width() == 1 );
  CPPUNIT_ASSERT( bmp2.Height() == 1 );
  CPPUNIT_ASSERT( bmp2.ColorTableSize() == 2 );

  // wrong bpp
  try {
    vp::Bmp bmp( 2, 3, 4 );
    CPPUNIT_ASSERT_MESSAGE( "Expected exception didn't throw", false );
  }
  catch( const vp::Exception& ) {}
}

void BmpTest::test1Bit()
{
  uint8_t B, G, R;
  vp::Bmp bmp( 1, 5, 5 );

  CPPUNIT_ASSERT( bmp.BitsPerPixel() == 1 );
  CPPUNIT_ASSERT( bmp.Width() == 5 );
  CPPUNIT_ASSERT( bmp.Height() == 5 );
  CPPUNIT_ASSERT( bmp.ColorTableSize() == 2 );

  // pixels
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 0 ); // 0 by default 
  }

  bmp.SetPixel( 0, 0, 1 );
  CPPUNIT_ASSERT( bmp.GetPixel( 0, 0 ) == 1 );
  bmp.SetPixel( 4, 4, 1 );
  CPPUNIT_ASSERT( bmp.GetPixel( 4, 4 ) == 1 );

  bmp.SetAllPixels( 1 );
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 1 );
  }

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( -1, 0, 1 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 5, 1 ), vp::Exception ); // y exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 2 ), vp::Exception );  // color index exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( -1, 0 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 5 ), vp::Exception ); // y exceeds

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 3, 4, 5 ), vp::Exception ); // not 24-bit
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 0, B, G, R ), vp::Exception ); // not 24-bit

  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( -1 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 2 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 3, 4, 5 ), vp::Exception ); // not 24-bit

  // color table
  for( uint8_t i = 0; i < bmp.ColorTableSize(); ++i )
  {
    bmp.GetColorTable( i, B, G, R );
    CPPUNIT_ASSERT( B == 0 );
    CPPUNIT_ASSERT( G == 0 );
    CPPUNIT_ASSERT( R == 0 );
  }

  bmp.SetColorTable( 1, 23, 24, 25 );
  bmp.GetColorTable( 1, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );

  CPPUNIT_ASSERT_THROW( bmp.SetColorTable( -1, 23, 24, 25 ), vp::Exception ); // color index
  CPPUNIT_ASSERT_THROW( bmp.SetColorTable( 2, 23, 24, 25 ), vp::Exception );  // color index
  CPPUNIT_ASSERT_THROW( bmp.GetColorTable( -1, B, G, R ), vp::Exception ); // color index
  CPPUNIT_ASSERT_THROW( bmp.GetColorTable( 2, B, G, R ), vp::Exception );  // color index
}

void BmpTest::test4Bits()
{
  uint8_t B, G, R;
  vp::Bmp bmp( 4, 8, 8 );

  CPPUNIT_ASSERT( bmp.BitsPerPixel() == 4 );
  CPPUNIT_ASSERT( bmp.Width() == 8 );
  CPPUNIT_ASSERT( bmp.Height() == 8 );
  CPPUNIT_ASSERT( bmp.ColorTableSize() == 16 );

  // pixels
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 0 ); // 0 by default 
  }

  bmp.SetPixel( 0, 0, 15 );
  CPPUNIT_ASSERT( bmp.GetPixel( 0, 0 ) == 15 );
  bmp.SetPixel( 7, 7, 15 );
  CPPUNIT_ASSERT( bmp.GetPixel( 7, 7 ) == 15 );

  bmp.SetAllPixels( 5 );
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 5 );
  }

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( -1, 0, 1 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 8, 1 ), vp::Exception ); // y exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 16 ), vp::Exception );  // color index exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( -1, 0 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 8 ), vp::Exception ); // y exceeds

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 3, 4, 5 ), vp::Exception ); // not 24-bit
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 0, B, G, R ), vp::Exception ); // not 24-bit

  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( -1 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 16 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 3, 4, 5 ), vp::Exception ); // not 24-bit

  // color table
  for( uint8_t i = 0; i < bmp.ColorTableSize(); ++i )
  {
    bmp.GetColorTable( i, B, G, R );
    CPPUNIT_ASSERT( B == 0 );
    CPPUNIT_ASSERT( G == 0 );
    CPPUNIT_ASSERT( R == 0 );
  }

  bmp.SetColorTable( 15, 23, 24, 25 );
  bmp.GetColorTable( 15, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );

  CPPUNIT_ASSERT_THROW( bmp.SetColorTable( -1, 23, 24, 25 ), vp::Exception ); // color index
  CPPUNIT_ASSERT_THROW( bmp.SetColorTable( 16, 23, 24, 25 ), vp::Exception );  // color index
  CPPUNIT_ASSERT_THROW( bmp.GetColorTable( -1, B, G, R ), vp::Exception ); // color index
  CPPUNIT_ASSERT_THROW( bmp.GetColorTable( 16, B, G, R ), vp::Exception );  // color index
}

void BmpTest::test8Bits()
{
  uint8_t B, G, R;
  vp::Bmp bmp( 8, 12, 12 );

  CPPUNIT_ASSERT( bmp.BitsPerPixel() == 8 );
  CPPUNIT_ASSERT( bmp.Width() == 12 );
  CPPUNIT_ASSERT( bmp.Height() == 12 );
  CPPUNIT_ASSERT( bmp.ColorTableSize() == 256 );

  // pixels
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 0 ); // 0 by default 
  }

  bmp.SetPixel( 0, 0, 255 );
  CPPUNIT_ASSERT( bmp.GetPixel( 0, 0 ) == 255 );
  bmp.SetPixel( 11, 11, 255 );
  CPPUNIT_ASSERT( bmp.GetPixel( 11, 11 ) == 255 );

  bmp.SetAllPixels( 5 );
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
      CPPUNIT_ASSERT( bmp.GetPixel( x, y ) == 5 );
  }

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( -1, 0, 1 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 12, 1 ), vp::Exception ); // y exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( -1, 0 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 12 ), vp::Exception ); // y exceeds

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 3, 4, 5 ), vp::Exception ); // not 24-bit
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 0, B, G, R ), vp::Exception ); // not 24-bit
  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 3, 4, 5 ), vp::Exception ); // not 24-bit

  // color table
  for( uint16_t i = 0; i < bmp.ColorTableSize(); ++i )
  {
    bmp.GetColorTable( i, B, G, R );
    CPPUNIT_ASSERT( B == 0 );
    CPPUNIT_ASSERT( G == 0 );
    CPPUNIT_ASSERT( R == 0 );
  }

  bmp.SetColorTable( 255, 23, 24, 25 );
  bmp.GetColorTable( 255, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );

  // caveat: -1 implicitly converted to 255
  bmp.SetColorTable( -1, 203, 204, 205 );
  bmp.GetColorTable( 255, B, G, R );
  CPPUNIT_ASSERT( B == 203 );
  CPPUNIT_ASSERT( G == 204 );
  CPPUNIT_ASSERT( R == 205 );

  // caveat: uint16_t 256 implicitly converted to 0
  bmp.SetColorTable( bmp.ColorTableSize(), 23, 24, 25 );
  bmp.GetColorTable( 0, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );
}

void BmpTest::test24Bits()
{
  uint8_t B, G, R;
  vp::Bmp bmp( 24, 12, 12 );

  CPPUNIT_ASSERT( bmp.BitsPerPixel() == 24 );
  CPPUNIT_ASSERT( bmp.Width() == 12 );
  CPPUNIT_ASSERT( bmp.Height() == 12 );
  CPPUNIT_ASSERT( bmp.ColorTableSize() == 0 );

  // pixels
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
    {
      bmp.GetPixel( x, y, B, G, R );
      CPPUNIT_ASSERT( B == 0 ); // 0 by default
      CPPUNIT_ASSERT( G == 0 );
      CPPUNIT_ASSERT( R == 0 );
    }
  }

  bmp.SetPixel( 0, 0, 23, 24, 25 );
  bmp.GetPixel( 0, 0, B, G, R );
  CPPUNIT_ASSERT( B == 23 );
  CPPUNIT_ASSERT( G == 24 );
  CPPUNIT_ASSERT( R == 25 );

  bmp.SetPixel( 11, 11, 203, 204, 205 );
  bmp.GetPixel( 11, 11, B, G, R );
  CPPUNIT_ASSERT( B == 203 );
  CPPUNIT_ASSERT( G == 204 );
  CPPUNIT_ASSERT( R == 205 );

  bmp.SetAllPixels( 50, 51, 52 );
  for( uint8_t y = 0; y < bmp.Width(); ++y )
  {
    for( uint8_t x = 0; x < bmp.Width(); ++x )
    {
      bmp.GetPixel( x, y, B, G, R );
      CPPUNIT_ASSERT( B == 50 );
      CPPUNIT_ASSERT( G == 51 );
      CPPUNIT_ASSERT( R == 52 );
    }
  }

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( -1, 0, 23, 24, 25 ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 12, 23, 24, 25 ), vp::Exception ); // y exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( -1, 0, B, G, R ), vp::Exception ); // x exceeds
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 12, B, G, R ), vp::Exception ); // y exceeds

  CPPUNIT_ASSERT_THROW( bmp.SetPixel( 0, 0, 255 ), vp::Exception ); // not indexed
  CPPUNIT_ASSERT_THROW( bmp.GetPixel( 0, 0 ), vp::Exception );     // not indexed

  CPPUNIT_ASSERT_THROW( bmp.SetAllPixels( 255 ), vp::Exception ); // not indexed

  // no color table
  CPPUNIT_ASSERT_THROW( bmp.SetColorTable( 0, 23, 24, 25 ), vp::Exception );
  CPPUNIT_ASSERT_THROW( bmp.GetColorTable( 0, B, G, R ), vp::Exception );
}

void BmpTest::testImport()
{
  vp::Bmp bmp;
  CPPUNIT_ASSERT( bmp.Import( "not_exist.bmp" ) == false );
  CPPUNIT_ASSERT_THROW( bmp.Import( "empty.bmp" ), vp::Exception );
}

void BmpTest::testExport()
{
  vp::Bmp bmp( 4, 10, 10 );

  // export to existing file
  CPPUNIT_ASSERT( bmp.Export( "export_exist.bmp" ) == false );
  CPPUNIT_ASSERT( bmp.Export( "export_exist.bmp", true ) );  // overwrite

  // export to new file
  CPPUNIT_ASSERT( bmp.Export( "export_new.bmp" ) );
}

File: test/py/BmpTest.py
#######################################################################
# Copyright (C) 2019 Xueyi Yao
#
# This file is part of VPixels.
#
# VPixels is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# VPixels is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with VPixels.  If not, see <https://www.gnu.org/licenses/>.
#######################################################################

# vp.bmp unit tests

import unittest
import sys

# import vpixels from current directory
sys_path = sys.path
if sys.platform == 'win32':
  sys.path = ['.\\']
else:
  sys.path = ['./']
import vpixels
sys.path = sys_path # restore default sys.path


class TestBmp( unittest.TestCase ):
  # vpixels.bmp() with no arg
  def testNoArg( self ):
    bmp = vpixels.bmp()
    self.assertEqual( 1, bmp.bitsperpixel() )
    self.assertEqual( 1, bmp.bpp() )
    self.assertEqual( 1, bmp.width() )
    self.assertEqual( 1, bmp.height() )
    self.assertEqual( (1, 1), bmp.dimension() )
    self.assertEqual( 2, bmp.colortablesize() )
    # default value 0
    self.assertEqual( (0, 0, 0), bmp.getcolortable( 0 ) )
    self.assertEqual( (0, 0, 0), bmp.getcolortable( 1 ) )
    bmp.setcolortable(0, 25, 26, 27)
    self.assertEqual( (25, 26, 27), bmp.getcolortable( 0 ) )
    # index exceeds limit
    self.assertRaises( ValueError, bmp.setcolortable, 2, 25, 26, 27 )
    self.assertRaises( ValueError, bmp.getcolortable, 2 )
    self.assertRaises( OverflowError, bmp.setcolortable, -1, 25, 26, 27 )
    self.assertRaises( OverflowError, bmp.getcolortable, -1 )


  # vpixels.bmp() with keywords
  def testKwords(self):
    bmp = vpixels.bmp( height = 7, bpp = 1, width = 9 )
    self.assertEqual( 1, bmp.bitsperpixel() )
    self.assertEqual( 1, bmp.bpp() )
    self.assertEqual( 9, bmp.width() )
    self.assertEqual( 7, bmp.height() )
    self.assertEqual( (9, 7), bmp.dimension() )
    self.assertEqual( 2, bmp.colortablesize() )

    bmp = vpixels.bmp( width = 7, bpp = 24, height = 9 )
    self.assertEqual( 24, bmp.bitsperpixel() )
    self.assertEqual( 24, bmp.bpp() )
    self.assertEqual( 7, bmp.width() )
    self.assertEqual( 9, bmp.height() )
    self.assertEqual( (7, 9), bmp.dimension() )
    self.assertEqual( 0, bmp.colortablesize() )


  # vpixels.bmp() with bad args
  def testBadArgs( self ):
    self.assertRaises( TypeError, vpixels.bmp, 2 )
    self.assertRaises( TypeError, vpixels.bmp, 2.5, 3, 4 )
    self.assertRaises( ValueError, vpixels.bmp, 2, 3, 4 )
    self.assertRaises( ValueError, vpixels.bmp, 9, 3, 4 )
    self.assertRaises( OverflowError, vpixels.bmp, -2, 3, 4 )
    self.assertRaises( ValueError, vpixels.bmp, 2, -3, 4 )
    self.assertRaises( ValueError, vpixels.bmp, 2, 3, -4 )
    self.assertRaises( TypeError, vpixels.bmp, 1, 3, 4, 5  )
    self.assertRaises( TypeError, vpixels.bmp, height = 5, width = 6 )
  

  # 1-bit
  def test1Bit( self ):
    bmp = vpixels.bmp( 1, 5, 6 )
    self.assertEqual( 1, bmp.bitsperpixel() )
    self.assertEqual( 1, bmp.bpp() )
    self.assertEqual( 5, bmp.width() )
    self.assertEqual( 6, bmp.height() )
    self.assertEqual( (5, 6), bmp.dimension() )

    self.assertEqual( 2, bmp.colortablesize() )
    bmp.setcolortable( 1, 2, 3, 4 )
    self.assertEqual( (2, 3, 4), bmp.getcolortable( 1 ) )
    self.assertRaises( ValueError, bmp.setcolortable, 2, 2, 3, 4 )
    self.assertRaises( OverflowError, bmp.setcolortable, -1, 2, 3, 4 )

    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    bmp.setallpixels( 1 )
    self.assertEqual( 1, bmp.getpixel( 0, 0 ) )
    bmp.setpixel( 0, 0, 0 )
    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    self.assertRaises( ValueError, bmp.getpixel, 5, 0 )
    self.assertRaises( ValueError, bmp.getpixel, 0, 6 )
    self.assertRaises( ValueError, bmp.setallpixels, 2 )
    self.assertRaises( ValueError, bmp.setpixel, 5, 0, 1 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 6, 1 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 0, 2 )

    bmp2 = bmp.clone()
    self.assertEqual( 1, bmp2.bitsperpixel() )
    self.assertEqual( 1, bmp2.bpp() )
    self.assertEqual( 5, bmp2.width() )
    self.assertEqual( 6, bmp2.height() )
    self.assertEqual( (5, 6), bmp2.dimension() )
    self.assertEqual( 2, bmp2.colortablesize() )
    self.assertEqual( (2, 3, 4), bmp2.getcolortable( 1 ) )
    self.assertEqual( 0, bmp2.getpixel( 0, 0 ) )
    self.assertEqual( 1, bmp2.getpixel( 1, 0 ) )
    

  # 4-bit
  def test4Bit( self ):
    bmp = vpixels.bmp( 4, 5, 6 )
    self.assertEqual( 4, bmp.bitsperpixel() )
    self.assertEqual( 4, bmp.bpp() )
    self.assertEqual( 5, bmp.width() )
    self.assertEqual( 6, bmp.height() )
    self.assertEqual( (5, 6), bmp.dimension() )

    self.assertEqual( 16, bmp.colortablesize() )
    bmp.setcolortable( 1, 2, 3, 4 )
    self.assertEqual( (2, 3, 4), bmp.getcolortable( 1 ) )
    self.assertRaises( ValueError, bmp.setcolortable, 16, 2, 3, 4 )
    self.assertRaises( OverflowError, bmp.setcolortable, -1, 2, 3, 4 )

    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    bmp.setallpixels( 5 )
    self.assertEqual( 5, bmp.getpixel( 0, 0 ) )
    bmp.setpixel( 0, 0, 0 )
    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    self.assertRaises( ValueError, bmp.getpixel, 5, 0 )
    self.assertRaises( ValueError, bmp.getpixel, 0, 6 )
    self.assertRaises( ValueError, bmp.setallpixels, 16 )
    self.assertRaises( ValueError, bmp.setpixel, 5, 0, 1 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 6, 1 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 0, 16 )

    bmp2 = bmp.clone()
    self.assertEqual( 4, bmp2.bitsperpixel() )
    self.assertEqual( 4, bmp2.bpp() )
    self.assertEqual( 5, bmp2.width() )
    self.assertEqual( 6, bmp2.height() )
    self.assertEqual( (5, 6), bmp2.dimension() )
    self.assertEqual( 16, bmp2.colortablesize() )
    self.assertEqual( (2, 3, 4), bmp2.getcolortable( 1 ) )
    self.assertEqual( 0, bmp2.getpixel( 0, 0 ) )
    self.assertEqual( 5, bmp2.getpixel( 1, 0 ) )


  # 8-bit
  def test8Bit( self ):
    bmp = vpixels.bmp( 8, 5, 6 )
    self.assertEqual( 8, bmp.bitsperpixel() )
    self.assertEqual( 8, bmp.bpp() )
    self.assertEqual( 5, bmp.width() )
    self.assertEqual( 6, bmp.height() )
    self.assertEqual( (5, 6), bmp.dimension() )

    self.assertEqual( 256, bmp.colortablesize() )
    bmp.setcolortable( 1, 2, 3, 4 )
    self.assertEqual( (2, 3, 4), bmp.getcolortable( 1 ) )
    self.assertRaises( OverflowError, bmp.setcolortable, 256, 2, 3, 4 )
    self.assertRaises( OverflowError, bmp.setcolortable, -1, 2, 3, 4 )

    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    bmp.setallpixels( 205 )
    self.assertEqual( 205, bmp.getpixel( 0, 0 ) )
    bmp.setpixel( 0, 0, 0 )
    self.assertEqual( 0, bmp.getpixel( 0, 0 ) )
    self.assertRaises( ValueError, bmp.getpixel, 5, 0 )
    self.assertRaises( ValueError, bmp.getpixel, 0, 6 )
    self.assertRaises( OverflowError, bmp.setallpixels, 256 )
    self.assertRaises( ValueError, bmp.setpixel, 5, 0, 1 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 6, 1 )
    self.assertRaises( OverflowError, bmp.setpixel, 0, 0, 256 )

    bmp2 = bmp.clone()
    self.assertEqual( 8, bmp2.bitsperpixel() )
    self.assertEqual( 8, bmp2.bpp() )
    self.assertEqual( 5, bmp2.width() )
    self.assertEqual( 6, bmp2.height() )
    self.assertEqual( (5, 6), bmp2.dimension() )
    self.assertEqual( 256, bmp2.colortablesize() )
    self.assertEqual( (2, 3, 4), bmp2.getcolortable( 1 ) )
    self.assertEqual( 0, bmp2.getpixel( 0, 0 ) )
    self.assertEqual( 205, bmp2.getpixel( 1, 0 ) )


  # 24-bit
  def test24Bit( self ):
    bmp = vpixels.bmp( 24, 5, 6 )
    self.assertEqual( 24, bmp.bitsperpixel() )
    self.assertEqual( 24, bmp.bpp() )
    self.assertEqual( 5, bmp.width() )
    self.assertEqual( 6, bmp.height() )
    self.assertEqual( (5, 6), bmp.dimension() )

    self.assertEqual( 0, bmp.colortablesize() )
    self.assertRaises( Exception, bmp.setcolortable, 1, 2, 3, 4 )
    self.assertRaises( Exception, bmp.getcolortable, 1 )

    self.assertEqual( (0, 0, 0), bmp.getpixel( 0, 0 ) )
    bmp.setallpixels( 205, 206, 207 )
    self.assertEqual( (205, 206, 207), bmp.getpixel( 0, 0 ) )
    bmp.setpixel( 0, 0, 25, 26, 27 )
    self.assertEqual( (25, 26, 27), bmp.getpixel( 0, 0 ) )
    self.assertRaises( ValueError, bmp.getpixel, 5, 0 )
    self.assertRaises( ValueError, bmp.getpixel, 0, 6 )
    self.assertRaises( TypeError, bmp.setallpixels, 25 )
    self.assertRaises( OverflowError, bmp.setallpixels, 256, 27, 28 )
    self.assertRaises( ValueError, bmp.setpixel, 5, 0, 25, 26, 27 )
    self.assertRaises( ValueError, bmp.setpixel, 0, 6, 25, 26, 27 )
    self.assertRaises( OverflowError, bmp.setpixel, 0, 0, 256, 27, 28 )
    self.assertRaises( OverflowError, bmp.setpixel, 0, 0, 26, 257, 28 )
    self.assertRaises( OverflowError, bmp.setpixel, 0, 0, 26, 27, 258 )

    bmp2 = bmp.clone()
    self.assertEqual( 24, bmp2.bitsperpixel() )
    self.assertEqual( 24, bmp2.bpp() )
    self.assertEqual( 5, bmp2.width() )
    self.assertEqual( 6, bmp2.height() )
    self.assertEqual( (5, 6), bmp2.dimension() )
    self.assertEqual( 0, bmp2.colortablesize() )
    self.assertEqual( (25, 26, 27), bmp2.getpixel( 0, 0 ) )
    self.assertEqual( (205, 206, 207), bmp2.getpixel( 1, 0 ) )


  def testImportf( self ):
    bmp = vpixels.bmp( 4, 5, 6 )

    # import non-existing file
    self.assertRaises( IOError, bmp.importf, 'not-exits.bmp' ) # not exit
    # bmp unchanged
    self.assertEqual( 4, bmp.bpp() )
    self.assertEqual( (5, 6), bmp.dimension() )
    self.assertEqual( 16, bmp.colortablesize() )

    # importf() expects a string argument
    self.assertRaises( TypeError, bmp.importf, 2 ) # not a file name
    # bmp unchanged
    self.assertEqual( 4, bmp.bpp() )
    self.assertEqual( (5, 6), bmp.dimension() )
    self.assertEqual( 16, bmp.colortablesize() )

    # import non-bmp file or corrupted bmp file
    self.assertRaises( Exception, bmp.importf, 'BmpTest.py' ) # not a bmp file
    # bmp becomes invalid, results of calling its methods are undefined


  def testExport( self ):
    bmp = vpixels.bmp()
    bmp.export( 'temp.bmp', True ) # overwrite = True
    self.assertRaises( IOError, bmp.export, 'temp.bmp' ) # overwrite = False
    self.assertRaises( IOError, bmp.export, 'temp.bmp', False ) # overwrite = False


  def testInheritance( self ):
    class subbmp( vpixels.bmp ):
      def __init__( self ):
        super(subbmp, self).__init__( 4, 3, 4 )

      def setcolor( self ):
        super(subbmp, self).setcolor( 0, 2, 3, 4 )

    bmp = subbmp()
    bmp.setcolor()
    self.assertEqual( (2, 3, 4), bmp.getcolor(0) )
    bmp.setpixel( 1, 2, 3)
    self.assertEqual( 3, bmp.getpixel( 1, 2 ) )


if __name__ == '__main__':
  unittest.main()
